source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,312 @@
+          rate_limit_settings.max_tokens_, dispatcher.timeSystem(), rate_limit_settings.fill_rate_);
+      drain_request_timer_ = dispatcher.createTimer([this]() -> void { drainRequests(); });
+    }
+    backoff_strategy_ =
+        std::make_unique<JitteredBackOffStrategy>(50,    // TODO RETRY_INITIAL_DELAY_MS,
+                                                  30000, // TODO RETRY_MAX_DELAY_MS,
+                                                  random);
+    establishNewStream();
+  }
+
+  // GRPC ACTUAL STREAM HANDLING STUFF
+  //  void start() { establishNewStream(); }
+
+  void setRetryTimer() {
+    retry_timer_->enableTimer(std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
+  }
+
+  void establishNewStream() {
+    ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
+    stream_ = async_client_->start(service_method_, *this);
+    if (stream_ == nullptr) {
+      ENVOY_LOG(warn, "Unable to establish new stream");
+      onIncrementalConfigFailed(nullptr);
+      setRetryTimer();
+      return;
+    }
+
+    // "must be populated for first request in a stream"
+    request_.clear_initial_resource_versions();
+    for (auto const& resource : resources_) {
+      (*request_.mutable_initial_resource_versions())[resource.first] = resource.second;
+    }
+
+    control_plane_stats_.connected_state_.set(1);
+    queueDiscoveryRequest();
+  }
+
+  void drainRequests() {
+    ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
+    while (!request_queue_.empty()) {
+      // Process the request, if rate limiting is not enabled at all or if it is under rate limit.
+      if (!rate_limiting_enabled_ || limit_request_->consume()) {
+        sendDiscoveryRequest();
+      } else {
+        ASSERT(rate_limiting_enabled_);
+        ASSERT(drain_request_timer_ != nullptr);
+        control_plane_stats_.rate_limit_enforced_.inc();
+        control_plane_stats_.pending_requests_.set(request_queue_.size());
+        drain_request_timer_->enableTimer(
+            std::chrono::milliseconds(limit_request_->nextTokenAvailableMs()));
+        break;
+      }
+    }
+  }
+
+  // Enqueues and attempts to send a discovery request with no change to subscribed resources.
+  void queueDiscoveryRequest() {
+    request_queue_.emplace();
+    drainRequests();
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void queueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    for (const auto& resource : resources) {
+      if (resources_.find(resource) == resources_.end()) {
+        diff.added_.push_back(resource);
+      }
+    }
+    for (const auto& entry : resources_) {
+      bool found = false;
+      for (const auto& resource : resources) {
```

Comment:
Should we use a set here to avoid O(n^2)?

Reply:
Heh, yup, will do.

I started out with a nice straightforward std::set_difference here back when I thought we could just work with sets of strings, rather than these added+removed diffs. I switched away when I introduced ResourceNameDiff because the necessary custom comparator thing was giving me trouble, and I didn't want the basic initial working implementation to get held up by that.

Reply:
Done.